contracts/modules/v2/Assessment.sol

```diff
@@ -0,0 +1,397 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+pragma solidity ^0.8.0;
+
+import "../../interfaces/INXMToken.sol";
+import "../../interfaces/ITokenController.sol";
+import "../../interfaces/IMemberRoles.sol";
+import "../../interfaces/IPool.sol";
+import "../../interfaces/IAssessment.sol";
+import "../../abstract/MasterAwareV2.sol";
+import "../../libraries/Assessment/AssessmentClaimsLib.sol";
+import "../../libraries/Assessment/AssessmentGovernanceActionsLib.sol";
+import "../../libraries/Assessment/AssessmentIncidentsLib.sol";
+import "../../libraries/Assessment/AssessmentUtilsLib.sol";
+import "../../libraries/Assessment/AssessmentVoteLib.sol";
+
+/**
+ *  Provides a way for cover owners to submit claims and redeem the payouts and facilitates
+ *  assessment processes where members decide the outcome of the events that lead to potential
+ *  payouts.
+ */
+contract Assessment is IAssessment, MasterAwareV2 {
+
+  /* ========== STATE VARIABLES ========== */
+
+  INXMToken internal nxm;
+
+  Configuration public CONFIG;
+
+  // ERC20 addresses of supported payout assetss (See Asset enum)
+  mapping(uint => address) internal addressOfAsset;
+
+  // Stake states of users. (See Stake struct)
+  mapping(address => Stake) public override stakeOf;
+
+  // Votes of users. (See Vote struct)
+  mapping(address => Vote[]) public override votesOf;
+
+  // Mapping used to determine if a user has already voted, using a vote hash as a key
+  mapping(bytes32 => bool) public override hasAlreadyVotedOn;
+
+  // An array of merkle tree roots used to indicate fraudulent assessors. Each root represents a
+  // fraud attempt by one or multiple addresses. Once the root is submitted by adivsory board
+  // members through governance, burnFraud uses this root to burn the fraudulent assessors' stakes
+  // and correct the outcome of the poll.
+  bytes32[] internal fraudMerkleRoots;
+  mapping(uint8 => mapping(uint104 => Poll)) internal pollFraudOfEvent;
+
+  Claim[] public override claims;
+
+  Incident[] public override incidents;
+  mapping(uint104 => address) internal incidentProponent;
+  mapping(uint104 => AffectedToken) internal tokenAffectedByIncident;
```

Comment:
why is it needed here? the product should know its token

Reply:
Three things:

- In the possibility where the product can be edited, the incident considers the token that was set at the time when it was proposed.
- Every time someone redeems a payout they don't have to make a call to the contract where the product details are stored, thus saving gas. There is no additional SSTORE for the proposer since we need priceBefore which is stored in the same slot anyway.
- It is also kept separately from incident as this information is only used at payout and in the UI. This means that we don't have an additional SLOAD everywhere else where we use the incident struct.